src/gui/torrentcontentmodel.cpp

```diff
@@ -41,27 +56,113 @@
 
 namespace
 {
+    const char showSystemFileIconsSettingKey[] = "Preferences/Advanced/showSystemFileIcons";
+
     QIcon getDirectoryIcon()
     {
         static QIcon cached = GuiIconProvider::instance()->getIcon("inode-directory");
         return cached;
     }
 
-    QIcon getFileIcon()
+    class UnifiedFileIconProvider: public QFileIconProvider
     {
-        static QIcon cached = GuiIconProvider::instance()->getIcon("text-plain");
-        return cached;
+    public:
+        using QFileIconProvider::icon;
+        QIcon icon(const QFileInfo &info) const override
+        {
+            Q_UNUSED(info);
+            static QIcon cached = GuiIconProvider::instance()->getIcon("text-plain");
+            return cached;
+        }
+    };
+#ifdef Q_OS_WIN
+    // See QTBUG-25319 for explanation why this is required
+    class WinShellFileIconProvider: public UnifiedFileIconProvider
+    {
+    public:
+        using QFileIconProvider::icon;
+        QIcon icon(const QFileInfo &info) const override
+        {
+            SHFILEINFO sfi = { 0 };
+            HRESULT hr = ::SHGetFileInfoW(info.absoluteFilePath().toStdWString().c_str(),
+                FILE_ATTRIBUTE_NORMAL, &sfi, sizeof(sfi), SHGFI_ICON | SHGFI_USEFILEATTRIBUTES);
+            if (FAILED(hr))
+                return UnifiedFileIconProvider::icon(info);
+#ifndef NO_QWINEXTRAS
+            QPixmap iconPixmap = QtWin::fromHICON(sfi.hIcon);
+#else
+            QPixmap iconPixmap = qt_pixmapFromWinHICON(sfi.hIcon);
+#endif
+            ::DestroyIcon(sfi.hIcon);
+            return QIcon(iconPixmap);
+        }
+    };
+#else
+    /**
+     * @brief Tests whether QFileIconProvider actually works
+     *
+     * Some QPA plugins do not implement QPlatformTheme::fileIcon(), and
+     * QFileIconProvider::icon() returns empty icons as the result. Here we asks it for
+     * two icons for probably absent files and when both icons are null, we assume that
+     * the current QPA plugin does not implement QPlatformTheme::fileIcon().
+     */
+    bool doesQFileIconProviderWork()
+    {
+        QFileIconProvider provider;
+        QIcon testIcon1 = provider.icon(QFileInfo(QLatin1String("/tmp/file.pdf")));
+        QIcon testIcon2 = provider.icon(QFileInfo(QLatin1String("/tmp/file.png")));
```

Comment:
would testing on more primitive file types better? e.g. `bmp` & `txt`.

>two icons for probably absent files

what happens if the file exist? or it doesn't matter?